Framework/NexusGeometry/src/InstrumentBuilder.cpp

```diff
@@ -0,0 +1,185 @@
+  verifyMutable();
+  if (!m_lastBank)
+    throw std::runtime_error("No bank to add the detector to");
+  auto *detector = new Geometry::Detector(
+      detName, detId,
+      const_cast<Geometry::IComponent *>(m_lastBank->getBaseComponent()));
+  detector->translate(Mantid::Kernel::toV3D(relativeOffset));
+  // No rotation set for detector pixels of a bank. This is not possible in the
+  // Nexus Geometry specification.
+  detector->setShape(shape);
+  m_lastBank->add(detector);
+  m_instrument->markAsDetectorIncomplete(detector);
+}
+
+/// Adds detector to instrument
+void InstrumentBuilder::addDetectorToInstrument(
+    const std::string &detName, detid_t detId, const Eigen::Vector3d &position,
+    boost::shared_ptr<const Geometry::IObject> &shape) {
+  verifyMutable();
+  auto *detector(new Geometry::Detector(
+      detName, detId,
+      const_cast<Geometry::IComponent *>(m_instrument->getBaseComponent())));
+  detector->setPos(position(0), position(1), position(2));
+
+  detector->setShape(shape);
+
+  m_instrument->add(detector);
+  m_instrument->markAsDetectorIncomplete(detector);
+}
+
+void InstrumentBuilder::addMonitor(
+    const std::string &detName, detid_t detId, const Eigen::Vector3d &position,
+    boost::shared_ptr<const Geometry::IObject> &shape) {
+  verifyMutable();
+  auto *detector(new Geometry::Detector(
+      detName, detId,
+      const_cast<Geometry::IComponent *>(m_instrument->getBaseComponent())));
+  detector->setPos(position(0), position(1), position(2));
+
+  detector->setShape(shape);
+
+  m_instrument->add(detector);
+  m_instrument->markAsMonitor(detector);
+}
+
+/// Sorts detectors
+void InstrumentBuilder::sortDetectors() const {
+  verifyMutable();
+  m_instrument->markAsDetectorFinalize();
+}
+
+/// Add sample
+void InstrumentBuilder::addSample(const std::string &sampleName,
+                                  const Eigen::Vector3d &position) {
+  verifyMutable();
+  auto *sample(this->addComponent(sampleName, position));
+  m_instrument->markAsSamplePos(sample);
+}
+/// Add source
+void InstrumentBuilder::addSource(const std::string &sourceName,
+                                  const Eigen::Vector3d &position) {
+  verifyMutable();
+  auto *source(this->addComponent(sourceName, position));
+  m_instrument->markAsSource(source);
+}
+
+void InstrumentBuilder::addBank(const std::string &localName,
+                                const Eigen::Vector3d &position,
+                                const Eigen::Quaterniond &rotation) {
+  verifyMutable();
+  auto *assembly =
+      new Geometry::CompAssembly(m_instrument->getBaseComponent(), nullptr);
+  assembly->setName(localName);
+  assembly->setPos(position[0], position[1], position[2]);
+  assembly->setRot(Kernel::toQuat(rotation));
+  m_lastBank = assembly;
+  m_instrument->add(assembly);
+}
+
+std::unique_ptr<const Geometry::Instrument>
+InstrumentBuilder::createInstrument() const {
+  verifyMutable();
+  sortDetectors();
+  m_finalized = true;
```

Comment:
Wouldn't it be simpler to remove the whole `verifyMutable` mechanism? Simply allocate a new empty instrument tree when `createInstrument()` is called, so the builder can be used again if required.